### PR TITLE
Correctly deal with time

### LINKED ISSUE #280: Deal correctly with time
Time brings a number of complications: 

* Calculating time from slot and vice versa can involve rounding errors
* Clock skew between nodes can cause nodes to reject blocks "in the future"

### PR DESCRIPTION
**Pull request by edsko**
The demo was rejecting chains because the new block was produced "in the
future"; except that here "in the future" was merely caused by clock skew. We
now prune blocks that are ahead by more than one block instead.

Closes #280.

### CODE REVIEW

File: ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime.hs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    .&&. slotStart <= now
+    .&&. 0 <= timeSpent
+    .&&. timeSpent < getSlotLength slotLen
+  where
+    now                      = addFixedUTC execTime (getSystemStart start)
+    (currentSlot, timeSpent) = slotAtTime slotLen start now
+    slotStart                = startOfSlot slotLen start currentSlot
+    slotLen'                 = getSlotLength slotLen
+
+-- > 0 < fst (timeUntilNextSlot now) <= slotLen
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime.hs]: I was wondering if the edge cases (close to 0 and `== slotLen`) were also being tested. Using `tabulate "Edge case:" [edgeCase]` and the code below, I got the following:

```haskell
    edgeCase
      | timeLeft < 3                      = "close to 0"
      | timeLeft == getSlotLength slotLen = "== slotLen"
      | otherwise                         = "in the middle"
```

    Edge case: (10000 in total):
    56.25% in the middle
    43.74% close to 0
     0.01% == slotLen

I had to run 10000 tests to get the 0.01%, but at least this "proves" it is correct.

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+-- | Slot length
+newtype SlotLength = SlotLength { getSlotLength :: FixedDiffTime }
+  deriving (Show)
+
+slotLengthFromMillisec :: Integer -> SlotLength
+slotLengthFromMillisec = SlotLength . FixedDiffTime . conv
+  where
+    conv :: Integer -> Fixed E3
+    conv = MkFixed
```
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: It's there to make sure _if_ we change that `E3` somewhere, then this code _must_ be adapted.

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+      , onSlotChange   = onEachChange id first (readTVar slotVar)
+      }
+
+{-------------------------------------------------------------------------------
+  Time utilities
+
+  We avoid using Double here to avoid rounding problems.
+-------------------------------------------------------------------------------}
+
+-- | Fixed precision UTC time (resolution: milliseconds)
```
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: Call it executive decision :joy:

File: ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime.hs (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime.hs]: Nice, John would be so proud :D I'll add this.

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: Will add a comment.
